### DISCUSSION TITLE
Request retransmissions or replays occur when a migration is finished?

### DISCUSSION THREAD

**Discussion by kingjunsung** [sarchlab/mgpusim#98] [Q&A] [answered]
Dear, Mr. Sun

When migration occurs, it appears that the command processor flushes the CU pipeline, cache, TLB, address translator, and RDMA, and then restarts the GPU. In this case, are the requests that did not receive responses from each component retransmitted or replayed?

If that’s not the case, then before migration occurs, it seems necessary to track all requests accessing the relevant page and ensure their responses. However, the simulator does not appear to have such a logic. Are the requests that did not receive responses being replayed entirely as mentioned above?

Thank you for your helps.

**Comment by syifan** [comment, marked as answer]
Replay only happens in the CU. A transaction that is not completed in the cache or TLB must maps to an uncompleted transaction in the CU, right? So, it is safe to only replay the transaction in the CU. 

These `shadowTransactions` are the transactions to replay. https://github.com/sarchlab/mgpusim/blob/v3/timing/cu/computeunit.go#L32
You can check how these fields are used to understand the replay mechanism.
